ltsp/server/ipxe/55-ipxe.sh

```diff
@@ -54,20 +59,36 @@ ipxe_main() {
         r_items="${r_items:+"$r_items\n"}$(printf "item --key %d %-20s %s" "$((key%10))" "r_$img_name" "$title")"
         r_gotos=":r_$img_name\nset img $img_name \&\& goto roots\n$r_gotos"
     done
+    m_items=""
+    m_gotos=":cmdline_boot_method"
+    img_name=$(re list_img_names -m)
+    set -- $img_name
+    for img_name in "$@"; do
+        key=$((key+1))
+        title=$(echo_values "$(ipxe_name "$img_name")")
+        title=${title:-$img_name}
+        m_items="${m_items:+"$m_items\n"}$(printf "item --key %d %-20s %s" "$((key%10))" "$img_name" "$title")"
+        m_gotos=":$img_name\nset img $img_name \&\& goto cmdline_boot_method\n$m_gotos"
+    done
     re mkdir -p "$TFTP_DIR/ltsp"
     if [ "$OVERWRITE" != "1" ] && [ -f "$TFTP_DIR/ltsp/ltsp.ipxe" ]; then
         warn "Configuration file already exists: $TFTP_DIR/ltsp/ltsp.ipxe
 To overwrite it, run: ltsp --overwrite $_APPLET ..."
     else
         client_sections=$(re client_sections)
-        re install_template "ltsp.ipxe" "$TFTP_DIR/ltsp/ltsp.ipxe" "\
+
+        sedp="\
 s|^/srv/ltsp|$BASE_DIR|g
 s/\(|| set menu-timeout \)5000/$(textif "$MENU_TIMEOUT" "\1$MENU_TIMEOUT" "&")/
 s|^:61:6c:6b:69:73:67\$|$(textif "$client_sections" "$client_sections" "&")|
-s|^#.*item.*\bimages\b.*|$(textif "$items$r_items" "$items\n$r_items" "&")|
-s|^:images\$|$(textif "$items" "$gotos" "&")|
+s|^#.*item.*\bimages\b.*|$(textif "$m_items$items$r_itmes" "$m_items\n$items\n$r_items" "&")|
 s|^:roots\$|$(textif "$r_items" "$r_gotos" "&")|
+s|^:images\$|$(textif "$items" "$gotos" "&")|
+s|^:cmdline_boot_method\$|$(textif "$m_items" "$m_gotos" "&")|
+s|\${CMDLINE_BOOT_METHOD}\$|$(textif "$CMDLINE_BOOT_METHOD" "$CMDLINE_BOOT_METHOD" "&")|
```

Comment:
So, I would only keep these, which would *replace* squashfs-over-nfs with squashfs-over-rbd. I wouldn't add more menu choices.

And other people could replace squashfs-over-nfs with squashfs-over-nbd, or squashfs-over-iscsi. We wouldn't want to generate all those menus just because we support multiple methods of booting.

Reply:
@alkisg, for clarification, my Ceph/RBD feature does not include squashfs. There are no images stored on a given Rados Block Device. Instead the respective Rados Block Device is the actual (partitioned) boot device, which can also be used to spin up a qemu/kvm virtual machine for doing maintenance on the image.

Also the snapshotting feature built into Rados allows admins to very conveniently maintain backup versions of their boot block devices!

So, it is not squashfs-over-rbd it is direct rbd boot.

However, I agree, there is a lot of repetition in the current 55-ipxe.sh code and portions could be merged in to more general patterns.

I don't completely understand what you are suggesting here. That you would put an IF statement in there and skip the :images and :roots, if CMDLINE_BOOT_METHOD is set? 

I thought about that also before, but I came to the conclusion, that even if you had RBD feature in use, you might still want to be able to have one of the other boot methods in the menu in parallel - for testing something quickly.

One thought I had however was, that :cmdline_boot_method actually supports all the other methods as well. So this is the most generalised target and could replace the other two. They would then just set their own cmdline_boot_method before they jump to this target.

NFS, iSCSI, NBD, FibreChannel or something completely different (e.g. HTTP) could then all be supported through the same :cmdline_boot_method.

Reply:
I can boot a VM disk file with LTSP like this:
mount flat-image.vmdk over NFS,
loop-mount it in /dev/loop8,
and then mount the first partition, /dev/loop8p1 to /root (the place where initramfs puts the root partition).

In an unrelated method, I can also call `ltsp image` to that file, and create a squashfs file, that is 2.5 times smaller and has less metadata, which then gives 2.5 times faster disk access.

And a third method is to tell LTSP to boot a disk and then get the squashfs image from inside it; this is used e.g. when booting a live-cd.iso with LTSP.

I do not know CephFS/RBD; help me understand. How would I boot a compressed squashfs file over CephFS, so that it's 2.5 times faster? Is CephFS the networking part, similar to NFS, so I could just use CephFS without RBD then?
Additionally, would I be able to partition and RBD device, and create a file system inside it, and put a squashfs file there, and ask LTSP to only boot that file?

Reply:
BTW RBDs would also support compression. No need for squashfs.

Reply:
RBD - Rados Block Device, as the name indicates is a block device, which is stored in a Rados storage cluster in a redundant distributed manner and provided to the system via network.

It is rather comparable to NBD than NFS.
So you can partition this block device just like an HDD and format the partitions or the whole block device with any FS.

Snapshotting and Compression is built into RBD, so there is no need to go down the squashfs (read-only) detour.

You can have anybody boot from the device RO or from a snapshot and have one boot RW, so you can install stuff or administer users. Maybe we could even figure out a way to make the others ssh-auth to this particular host - just some thoughts anyways.

CephFS on the other hand is a fully POSIX compliant network file system, ergo more like NFS (which however is NOT fully POSIX compliant, which e.g. causes problems with Gnome dconf).

Every user could even have their own ceph access key, provided to them on ssh-auth and allowing them only to mount /home/user instead of all /home. That is the next part I will work on.

It is beautiful.